File: Java/1427. Perform String Shifts.java
class Solution {
    public String stringShift(String s, int[][] shift) {
        int S = s.length(), shifts = 0;

        for (int[] ss: shift) {
            if (ss[0] == 1) {
                ss[1] *= -1;
            }
            shifts += ss[1];
        }

        shifts = Math.floorMod(shifts, S);
        s = s.substring(shifts) + s.substring(0, shifts);
        return s;
    }
}

File: Python3/1427. Perform String Shifts.py
class Solution:
    def stringShift(self, s: str, shift: List[List[int]]) -> str:
        shifts = 0
        S = len(s)

        for direction, amount in shift:
            # left: increase
            if direction == 0:
                shifts += amount

            # right: decrease
            else:
                shifts -= amount

        shifts %= S
        # shift left
        if shifts > 0:
            s = s[shifts:] + s[:shifts]
        elif shifts < 0:
            s = s[S + shifts:] + s[:S + shifts]

        return s